L74-java-channel-creds.md

```diff
@@ -0,0 +1,571 @@
+      for (ChannelCredentials innerCreds : choiceCreds.getCredentialsList()) {
+        Result result = from(innerCreds);
+        if (!result.isError()) {
+          return result;
+        }
+        error.append("; ");
+        error.append(result.getError());
+      }
+      return Result.error(error.substring(2));
+
+    } else {
+      return Result.error(
+          "Unsupported credential type: " + creds.getClass().getName());
+    }
+  }
+
+  public static final class Result {
+    public static Result error(String error) {...}
+    public static Result negotiator(ProtocolNegotiator.Factory factory) {...}
+    public Result withCallCredentials(CallCredentials callCreds) {...}
+  }
+}
+```
+
+### ServerCredentials
+
+Server-side will be a clone of client-side, but using the `Server` instead of
+`Channel`. `ServerCredentials` cannot contain `CallCredentials`, so
+`CompositeServerCredentials` will be omitted.
+
+```java
+// instead of:
+server = ServerBuilder.forPort(443)
+    .useTransportSecurity(new File("cert.pem"), new File("cert.key"))
+    .build().start();
+server = ServerBuilder.forPort(8080)
+    .build().start();
+// the user would now:
+ServerCredentials tlsCreds = new TlsServerCredentials(
+    new File("cert.pem"), new File("cert.key"));
+server = Grpc.newServerBuilderForPort(443, tlsCreds)
+    .build().start();
+server = Grpc.newServerBuilderForPort(8080, new InsecureChannelCredentials())
+    .build().start();
+```
+
+
+```java
+package io.grpc;
+
+public abstract class ServerCredentials {}
+```
+
+Since servers using TLS require a certificate, the `TlsServerCredentials`
+cannot have the simple no-arg configuration of `TlsChannelCredentials`. It will
+provide convenience constructors for providing a certificate and unencrypted
+key. If any other configuration is necessary, the `Builder` would be used.
+
+```java
+package io.grpc;
+
+public final class TlsServerCredentials extends ServerCredentials {
+  private TlsServerCredentials(...) {}
+
+  /**
+   * Creates an instance using provided certificate chain and private key.
+   * Generally they should be PEM-encoded and the key is an unencrypted PKCS#8
+   * key (file headers have "BEGIN CERTIFICATE" and "BEGIN PRIVATE KEY").
+   */
+  public static ServerCredentials create(File certChain, File privateKey)
+      throws IOException {
+    return newBuilder().keyManager(certChain, privateKey).build();
+  }
+  // Will not auto-close InputStream, unlike existing transportSecurity()
+  public static ServerCredentials create(
+      InputStream certChain, InputStream privateKey) throws IOException {
+    return newBuilder().keyManager(certChain, privateKey).build();
+  }
+
+  public static Builder newBuilder() { return new Builder(); }
+
+  /** Same as client-side */
+  public Set<Feature> incomprehensible(Set<Feature> understoodFeatures) {...}
+
+  public byte[] getCertificateChain() { return Arrays.copyOf(...); }
+  public byte[] getPrivateKey() { return Arrays.copyOf(...); }
```

Comment:
Couple of comments:
- shouldn't the same or similar methods be available in `TlsChannelCredentials` to get the identity and root certificates? I didn't see them defined above. The transport layers will need these methods right?
- for the mTLS case , `getRootCertificate` may also need to be defined

Reply:
Client-side: eventually they will be added as added Features, but the basic functionality doesn't require it. Today there is not a stable API to specify the identity and root certs client-side.

Server-side: ditto, really. mTLS is not supported by the pre-existing stable TLS APIs. You have to use transport-specific APIs. Eventually that will likely be added, but not as a first cut.

Reply:
Yes, makes sense. I keep thinking of mTLS as basic functionality :-) On the client side - with just TLS - I understand there is a default trust-store.